File: hs/day3.hs
import Control.Exception(assert)
import Data.ByteString as BS(ByteString, length, readFile, splitAt, unpack)
import qualified Data.ByteString.Char8 as C
import Data.List(foldl1')
import Data.List.Split
import Data.Word
import Data.Set

-- Priority (score) of a given item, [a-z,A-Z]
prioOf :: Word8 -> Word64
prioOf c = fromIntegral $ prioOf' $ fromIntegral c -- Just casts to int
    where
        prioOf' :: Int -> Int
        prioOf' i
            | i >= fromEnum 'a' && i <= fromEnum 'z' = i - fromEnum 'a' + 1
            | i >= fromEnum 'A' && i <= fromEnum 'Z' = i - fromEnum 'A' + 27
            | otherwise =  error "wat"


part1 :: ByteString -> Word64
part1 input = sum $ duplicatedItemScore <$> C.lines input

-- Convert a ByteString to a set of its bytes
toSet :: ByteString -> Set Word8
toSet = fromList . BS.unpack

-- Grab the element out of a single-elemnt set
onlyElement :: Set a -> a
onlyElement s = assert (size s == 1) $ head . toList $ s

-- Find the (score of the) duplicate item in both sides of the pack.
duplicatedItemScore :: ByteString -> Word64
duplicatedItemScore line = prioOf duped
    where
        half = BS.length line `div` 2 :: Int
        (left, right) = BS.splitAt half line
        (leftSet, rightSet) = (toSet left, toSet right)
        duped = onlyElement $ intersection leftSet rightSet

part2 :: ByteString -> Word64
part2 input = sum $ groupScore <$> chunksOf 3 (C.lines input)

-- Find the (score of the) common item in all three packs.
groupScore :: [ByteString] -> Word64
groupScore group = prioOf . onlyElement $ shared
    where
        sets = toSet <$> group
        shared = foldl1' intersection sets

main :: IO ()
main = do
    input <- BS.readFile "../input/2022/day3.txt"
    print $ part1 input
    print $ part2 input

File: src/day3.rs
use aoc_runner_derive::aoc;

use itertools::Itertools;
use rustc_hash::FxHashSet;

fn calc_prio(item: u8) -> i64 {
    if (b'a'..=b'z').contains(&item) {
        (item - b'a') as i64 + 1
    } else if (b'A'..=b'Z').contains(&item) {
        (item - b'A') as i64 + 27
    } else {
        unreachable!("bad input");
    }
}

#[aoc(day3, part1)]
fn part1(input: &str) -> i64 {
    let mut prios = 0i64;

    for line in input.lines().map(|l| l.as_bytes()) {
        let half = line.len() / 2;
        let sack1 = to_set(&line[..half]);
        let sack2 = to_set(&line[half..]);

        let in_both = sack1.intersection(&sack2).collect::<Vec<_>>();
        assert_eq!(in_both.len(), 1);

        let in_both = *in_both[0];

        prios += calc_prio(in_both);
    }
    prios
}

fn to_set(l: &[u8]) -> FxHashSet<u8> {
    l.iter().copied().collect()
}

#[aoc(day3, part2)]
fn part2(input: &str) -> i64 {
    let mut prios = 0i64;
    for group in input
        .lines()
        .map(|l| to_set(l.as_bytes()))
        .chunks(3)
        .into_iter()
    {
        let shared = group
            .reduce(|a, b| a.intersection(&b).copied().collect())
            .expect("empty group");
        assert_eq!(shared.len(), 1);
        prios += calc_prio(*shared.iter().next().unwrap());
    }
    prios
}
